### PR TITLE
Update HttpInListener to add gRPC tags - By creating a new activity with the OperationName used by the framework

### LINKED ISSUE #1740: Instrumentation.AspNetCore for gRPC services omits ALL rpc.* attributes under certain conditions
# Bug Report

NuGet package(s):
- OpenTelemetry.Instrumentation.AspNetCore 1.0.0-rc.1 NuGet package (and tip of tree)

Runtimes:
- netcoreapp3.1

## Symptom

rpc.* attributes required for RPC Spans are not added to the created gRPC server Span under particular circumstances. All I see are http.* attributes. This happens when using a composite text map propagator made up of B3 and W3C in that order (B3 takes precedence if there). We have nginx and linkerd services sitting in front of these netcore gRPC services which understand and augment traces via the B3 headers before passing through to the gRPC service. Existing traceparent headers are treated as opaque data and forwarded unmodified as intended. The gRPC instrumentation correctly extracts the context from the B3 headers but still fails to output the rpc attributes.

This is happening because a new Activity is created in this block: https://github.com/open-telemetry/opentelemetry-dotnet/blob/ccc191f0c61a5a16c5a4cf577f82a7ac1c5dbd41/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L86 with a name (ActivityCreatedByHttpInListener) that doesn't satisfy a check in the dotnet-grpc code which is responsible for adding a tag on the Activity called grpc.method. It looks specifically for an Activity named Microsoft.AspNetCore.Hosting.HttpRequestIn. See https://github.com/grpc/grpc-dotnet/blob/76ae64d3b0e28772bc4d97c0d8a146b3609b465d/src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs#L410

The otel instrumentation looks for the presence of that grpc.method tag before adding the rpc.* tags.

**What is the expected behavior?**

rpc.* attributes are present according to https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/rpc.md

## Repro
Add a composite propagator to the Examples.Grpc.Service. See the last commit on my test branch: https://github.com/pcwiese/opentelemetry-dotnet/commit/6cc33a137fad61ff852c9dabc47cf39a2722c410

Start the Examples.Grpc.Service and hit it with a grpcurl request similar to this one:

```
grpcurl.exe -proto "Z:\src\opentelemetry-dotnet\examples\GrpcService\Protos\greet.proto" -import-path "Z:\src\opentelemetry-dotnet\examples\GrpcService\Protos" -H "traceparent: 00-120dc44db5b736468afb112197b0dbd3-5dfbdf27ec544544-01" -H "x-b3-traceid: 120dc44db5b736468afb112197b0dbd3" -H "x-b3-spanid: b0966f651b9e0126" -H "x-b3-sampled: 1" localhost:44335 greet.Greeter/SayHello
```

The trace ids match between the traceparent header and B3 headers but the span ids do not. When you make the request you should see the rpc.* attributes outputted from the service but they aren't there.

## Possible fix
Given that the Grpc.AspNetCore.Server NuGet package **requires** the Activity to be a certain name, we could simply create the Activity as we do now in the instrumentation library, but use the incoming Activity's OperationName instead of ActivityCreatedByHttpInListener. I tried this out locally and it does work but I'm sure what the implication of that is elsewhere.

### PR DESCRIPTION
**Pull request by utpilla**
Fixes #1740 

## Changes
- Updated `OnStartActivity` method to create a sibling activity using the same OperationName as that of the activity created by framework. This leads to the framework adding the `grpc.method` and `grpc.status_code` tags to the activity created by the instrumentation
- ~~Added a boolean custom property called `IsCreatedByInstrumentation` to know if the activity was created by instrumentation~~
- Add a tag `IsCreatedByInstrumentation` and set it to `bool.TrueString` right after creating the activity in the listener so that we only have to check the first tag to determine if the activity was created by instrumentation
- Stop the activity created by instrumentation in the `OnStopActivity` method by checking the newly added tag `IsCreatedByInstrumentation`
- Updated `AddAspNetCoreInstrumentation` extension method to not add `ActivityCreatedByHttpInListener` as a legacy source
- Added a unit test for the gRPC scenario where the instrumentation creates a new activity
- Added an ActivityExtension method named `TryCheckFirstTag`

- [x] Updated CHANGELOG.md

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs
Comment (pcwiese) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:216]: The if block above this would never be true if we reuse the operation name right?
`if (activity.OperationName.Equals(ActivityNameByHttpInListener, StringComparison.Ordinal))
            {
`

Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:216]: Yeah, if we decide to go forward with this approach, we can get rid of the above if block.

File: test/OpenTelemetry.Instrumentation.Grpc.Tests/GrpcTests.server.cs
Comment (cijothomas) [inline test/OpenTelemetry.Instrumentation.Grpc.Tests/GrpcTests.server.cs:187]: do this inside a finally block.

Comment (utpilla) [reply test/OpenTelemetry.Instrumentation.Grpc.Tests/GrpcTests.server.cs]: Fixed this.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs
Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:101]: This is not a common scenario. If we use a tag to keep track of this, we will have to remove this tag before activity.Stop so that this tag does not get exported. This would mean that we will have to look through the TagObjects for this tag for all scenarios. The most commonly occurring scenario would be where there is no new activity created and this tag does not get added. We will have to iterate through all the TagObjects to confirm that no new activity was created which might be a little more expensive than looking up a custom property.

I feel the customProperty approach is good to unblock and fix this issue. We need to run benchmark tests for customProperties vs TagObjects approach to find the most performant one. I have created an issue #1887 to track this

Comment (CodeBlanch) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:101]: If the user is using some form of propagation other than W3C will it fire 100% of the time? Rare I think would be like it only happens for connection failures.

Some other (maybe crazy) ideas I came up with trying to think outside the box:

* The AspNetCore created Activity never has a parent? What if we set a parent on the one we create and use that as the signal?

* AspNetCore created Activity doesn't have an ActivitySource set? What if we set one to use as the signal? It is a `private set` so we would need a bit of magic for that.

**Comment by pcwiese** [comment]
Is there something blocking this PR?

**Comment by cijothomas** [comment]
Assigned to Grpc expert @alanwest for further review.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs
Comment (alanwest) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:101]: > If the user is using some form of propagation other than W3C will it fire 100% of the time?

Yes, is true.

> The AspNetCore created Activity never has a parent?

I do not think this is true. AspNetCore created activity will have a parent if the calling service included W3C context headers in the request.

> AspNetCore created Activity doesn't have an ActivitySource set? What if we set one to use as the signal? It is a private set so we would need a bit of magic for that.

This seems like a legit solution. AspNetCore may eventually get instrumented with ActivitySource, but then I guess this instrumentation would not be invoked, so we're probably good.

**Review by alanwest — COMMENTED** [review]
@utpilla The fix seems good. I agree with @CodeBlanch we should consider perf implications. I think the `ActivitySource` check has some good potential.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs
Comment (eddynaka) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:101]: i think u could replace for bool.True (something like this)

Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs]: Fixed this.

**Review by CodeBlanch — APPROVED** [review]
LGTM

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:101]: Can we do `SetTag` above the Start() call? This will ensure that this will be the 1st tag. (Because someone could setup a listener to the Start() callback and add a new tag there, which will affect our assumption)

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:218]: should we just remove this tag here? (if you set a tag value to null, that would be as good as removing it). Otherwise an exporter sees this as just any other tag. We don't want to leak the implementation details there.

**Review by cijothomas — CHANGES_REQUESTED** [review]
LGTM. 
Blocking to address couple of comments. Please address them too and also add changelog entry.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs
Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:101]: This tag is being added to the activity that we create in our callback logic. So even if someone adds a listener to DiagnosticSource start, they will not have access to the activity that we create in our callback method, right?

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:101]: I meant ActivityListener. Anyone can listen to *any* activities, including the ones we create here :)

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs
Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs]: Fixed this.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs
Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs]: Yeah makes sense. Fixed this.

**Review by cijothomas — APPROVED** [review]

**Review by cijothomas — APPROVED** [review]

**Pull request merged**
